File: tests/legacy/server/chat-process.ts
import type { ChatExecution, ChatMessage } from '../../../shared/chats.ts'
import type { Config } from './config.ts'
import { createHash } from 'node:crypto'
import { readFile, writeFile } from 'node:fs/promises'
import process from 'node:process'
import { questionFields } from '../../../shared/chats.ts'
import { codexSession } from './codex-rpc.ts'

export interface ChatPlan {
  execution: ChatExecution
  sessionId?: string
  instructions: string
  inputDirectory: string
  output: string
  cwd: string
  model: string
  reasoning: string
  sandbox: 'yolo' | 'workspace-write' | 'read-only'
  writableRoots: string[]
  args: string[]
}
const emit = (event: Record<string, unknown>) => process.stdout.write(`${JSON.stringify(event)}\n`)
const input = (text: string) => [{ type: 'text', text, text_elements: [] }]
// Keep the established Activity artifact contract independent of Codex's transport.
export function chatItem(item: Record<string, any>) {
  const types: Record<string, string> = { agentMessage: 'agent_message', commandExecution: 'command_execution', fileChange: 'file_change', mcpToolCall: 'mcp_tool_call', reasoning: 'reasoning', plan: 'reasoning', webSearch: 'web_search', collabAgentToolCall: 'collab_tool_call' }
  return { ...item, type: types[item.type] ?? item.type, aggregated_output: item.aggregatedOutput, exit_code: item.exitCode, duration_ms: item.durationMs, text: item.text ?? item.summary?.join('\n'), changes: item.changes?.map((change: any) => ({ ...change, kind: typeof change.kind === 'object' ? change.kind.type : change.kind })) }
}
export async function runChat(plan: ChatPlan, binary = 'codex') {
  let threadId = ''
  let turnId = ''
  let finished = false
  let lastMessage = ''
  let settle!: (turn: any) => void
  let rejectCompletion!: (error: Error) => void
  const completed = new Promise<any>((resolve, reject) => {
    settle = resolve
    rejectCompletion = reject
  })
  void completed.catch(() => {})
  const seen = new Set<string>()
  const attempted = new Set<string>()
  const texts = new Map<string, string>()
  const questions = new Map<string, { requestId: string | number, reply: (result: unknown) => void, messageId?: string }>()
  const questionId = (itemId: string) => createHash('sha256').update(`${threadId}:${itemId}`).digest('hex')
  const acknowledge = (id: string) => {
    if (seen.has(id))
      return
    seen.add(id)
    emit({ type: 'chat.delivered', messageId: id })
  }
  const onItem = (item: any, type = 'item.completed') => {
    if (item.type === 'functionCallOutput' && ['request_user_input', 'request_user_input_async'].includes(item.name))
      return
    if (item.type === 'userMessage') {
      if (item.clientId)
        acknowledge(item.clientId)
      return
    }
    if (item.type === 'agentMessage') {
      lastMessage = item.text || lastMessage
      if (type === 'item.completed' && item.questions?.length) {
        const fields = questionFields.safeParse(item.questions.map((question: any, index: number) => ({ id: `${index}`, title: question.title, options: question.options?.map((label: string) => ({ label })) ?? [] })))
        if (fields.success)
          emit({ type: 'chat.question', question: { id: questionId(item.id), blocking: false, fields: fields.data } })
      }
    }
    emit({ type, item: chatItem(item) })
  }
  const session = codexSession({ codexBin: binary, home: process.env.HOME } as Config, {
    args: plan.args,
    closed: () => rejectCompletion(new Error('Codex disconnected before finishing the response. Resume the conversation to continue.')),
    cwd: plan.cwd,
    serverRequest(method, params, reply, requestId) {
      if (method !== 'item/tool/requestUserInput' || !params?.itemId || (threadId && params.threadId !== threadId))
        return false
      const fields = questionFields.safeParse(params.questions?.map((question: any) => ({ id: question.id, title: question.question, secret: question.isSecret ?? false, options: question.options ?? [] })))
      if (!fields.success)
        return false
      const id = questionId(params.itemId)
      questions.set(id, { requestId, reply })
      emit({ type: 'chat.question', question: { id, blocking: params.isBlocking !== false, fields: fields.data } })
      return true
    },
    notification(method, params) {
      if (params?.threadId && threadId && params.threadId !== threadId)
        return
      if (method === 'serverRequest/resolved') {
        for (const [id, question] of questions) {
          if (question.requestId !== params.requestId)
            continue
          if (question.messageId)
            acknowledge(question.messageId)
          questions.delete(id)
          emit({ type: 'chat.question.closed', questionId: id })
        }
      }
      if (method === 'turn/started') {
        turnId = params.turn.id
        emit({ type: 'turn.started' })
      }
      if (method === 'item/started' || method === 'item/completed')
        onItem(params.item, method.replace('/', '.'))
      if (method === 'item/agentMessage/delta') {
        const text = (texts.get(params.itemId) ?? '') + params.delta
        texts.set(params.itemId, text)
        emit({ type: 'item.updated', item: { id: params.itemId, type: 'agent_message', text } })
      }
      if (method === 'turn/completed') {
        finished = true
        settle(params.turn)
      }
    },
  })
  await session(process.env.CODEX_HOME!, async (rpc) => {
    const settings = { cwd: plan.cwd, model: plan.model || undefined, approvalPolicy: 'never', sandbox: plan.sandbox === 'yolo' ? 'danger-full-access' : plan.sandbox, developerInstructions: plan.instructions, config: { 'features.default_mode_request_user_input': true, 'model_reasoning_effort': plan.reasoning, 'sandbox_workspace_write': { network_access: true, writable_roots: plan.writableRoots } } }
    const result = await rpc.request<any>(plan.sessionId ? 'thread/resume' : 'thread/start', { ...settings, ...(plan.sessionId ? { threadId: plan.sessionId, excludeTurns: false } : {}) })
    threadId = result.thread.id
    emit({ type: 'chat.question.closed' })
    emit({ type: 'thread.started', thread_id: threadId })
    // Client IDs survive in the Codex transcript: a crash after acceptance must
    // recover the existing turn, not submit the user's instruction twice.
    const turns = result.thread.turns ?? []
    if (plan.sessionId && result.thread.historyMode === 'paginated') {
      turns.length = 0
      let cursor: string | undefined
      do {
        const page = await rpc.request<any>('thread/turns/list', { threadId, cursor, limit: 100, itemsView: 'full', sortDirection: 'desc' })
        turns.push(...page.data)
        cursor = page.nextCursor ?? undefined
      } while (cursor)
    }
    for (const turn of turns) {
      for (const item of turn.items ?? []) {
        if (item.type === 'userMessage' && item.clientId)
          acknowledge(item.clientId)
      }
    }
    const acceptedTurn = turns.findLast((turn: any) => turn.items?.some((item: any) => item.clientId === plan.execution.messageId))
    const previous = acceptedTurn && plan.execution.recovery ? (result.thread.historyMode === 'paginated' ? turns[0] : turns.at(-1)) : acceptedTurn
    if (previous) {
      emit({ type: 'chat.delivered', messageId: plan.execution.messageId })
      if (previous.status === 'completed') {
        for (const item of previous.items) onItem(item)
        await writeFile(plan.output, lastMessage)
        emit({ type: 'turn.completed' })
        return
      }
    }
    const recovering = !!previous
    const start = await rpc.request<any>('turn/start', {
      threadId,
      clientUserMessageId: recovering ? undefined : plan.execution.messageId,
      input: input(recovering ? `Continue the interrupted conversation from its last completed step. Preserve completed work and verify external effects before repeating any action. The pending user request is:\n${plan.execution.text}` : plan.execution.text),
      model: plan.model || undefined,
      effort: plan.reasoning,
    })
    turnId = start.turn.id
    acknowledge(plan.execution.messageId)
    let polling = false
    const timer = setInterval(async () => {
      if (polling || finished || !turnId)
        return
      polling = true
      try {
        const messages: ChatMessage[] = JSON.parse(await readFile(`${plan.inputDirectory}/messages.json`, 'utf8'))
        for (const message of messages) {
          if (finished || seen.has(message.id) || attempted.has(message.id))
            continue
          attempted.add(message.id)
          try {
            const question = message.questionId ? questions.get(message.questionId) : undefined
            if (question && message.answers) {
              question.messageId = message.id
              question.reply({ answers: Object.fromEntries(Object.entries(message.answers).map(([id, answers]) => [id, { answers }])) })
              continue
            }
            await rpc.request('turn/steer', { threadId, expectedTurnId: turnId, clientUserMessageId: message.id, input: input(message.text) })
            acknowledge(message.id)
          }
          catch {
            // The turn may have ended between the click and the request. Leave
            // the message queued for the next turn; never discard it.
            break
          }
        }
      }
      catch { /* The worker has not published an inbox yet. */ }
      finally { polling = false }
    }, 250)
    try {
      const turn = await completed
      clearInterval(timer)
      // Polling is cleared by the in-flight steering request.
      // eslint-disable-next-line no-unmodified-loop-condition
      while (polling) await new Promise(resolve => setTimeout(resolve, 10))
      if (turn.status !== 'completed') {
        emit({ type: 'turn.failed', error: turn.error ?? { message: 'Conversation interrupted.' } })
        process.exitCode = 1
        return
      }
      await writeFile(plan.output, lastMessage)
      emit({ type: 'turn.completed' })
    }
    finally { clearInterval(timer) }
  })
}
if (process.argv[1] === import.meta.filename) {
  let data = ''
  for await (const chunk of process.stdin) data += chunk
  runChat(JSON.parse(data), process.argv[2]).catch((error) => {
    emit({ type: 'turn.failed', error: { message: error.message } })
    process.exitCode = 1
  })
}

File: backend/src/chat_process.rs
use crate::{
    auth::hex_digest,
    config::Config,
    error::{Error, Result},
    rpc::{Incoming, Session},
    skills::atomic_write,
    validation::{parse, text},
};
use serde_json::{Value, json};
use std::{
    collections::{HashMap, HashSet},
    path::Path,
    time::Duration,
};
use tokio::sync::mpsc;
use tokio_util::sync::CancellationToken;
pub fn chat_item(mut item: Value) -> Value {
    let kind = match text(&item, "type") {
        "agentMessage" => "agent_message",
        "commandExecution" => "command_execution",
        "fileChange" => "file_change",
        "mcpToolCall" => "mcp_tool_call",
        "reasoning" | "plan" => "reasoning",
        "webSearch" => "web_search",
        "collabAgentToolCall" => "collab_tool_call",
        kind => kind,
    }
    .to_owned();
    item["type"] = kind.into();
    for (source, target) in [
        ("aggregatedOutput", "aggregated_output"),
        ("exitCode", "exit_code"),
        ("durationMs", "duration_ms"),
    ] {
        if let Some(value) = item.get(source).cloned() {
            item[target] = value;
        }
    }
    if item["text"].is_null()
        && let Some(summary) = item["summary"].as_array()
    {
        item["text"] = summary
            .iter()
            .filter_map(Value::as_str)
            .collect::<Vec<_>>()
            .join("\n")
            .into();
    }
    if let Some(changes) = item["changes"].as_array_mut() {
        for change in changes {
            if change["kind"].is_object() {
                change["kind"] = change["kind"]["type"].clone();
            }
        }
    }
    item
}
struct Question {
    request_id: Value,
    message_id: Option<String>,
}
struct Chat {
    session: Session,
    events: mpsc::Sender<Value>,
    cancel: CancellationToken,
    thread: String,
    turn: String,
    completed: Option<Value>,
    last_message: String,
    seen: HashSet<String>,
    attempted: HashSet<String>,
    texts: HashMap<String, String>,
    questions: HashMap<String, Question>,
}
impl Chat {
    async fn emit(&self, value: Value) -> Result<()> {
        self.events
            .send(value)
            .await
            .map_err(|_| Error::new(503, "Conversation output stopped."))
    }
    async fn acknowledge(&mut self, id: &str) -> Result<()> {
        if self.seen.insert(id.to_owned()) {
            self.emit(json!({
            "type":"chat.delivered","messageId":id}
            ))
            .await?;
        }
        Ok(())
    }
    async fn item(&mut self, item: Value, kind: &str) -> Result<()> {
        if item["type"] == "functionCallOutput"
            && ["request_user_input", "request_user_input_async"].contains(&text(&item, "name"))
        {
            return Ok(());
        }
        if item["type"] == "userMessage" {
            if let Some(id) = item["clientId"].as_str() {
                self.acknowledge(id).await?;
            }
            return Ok(());
        }
        if item["type"] == "agentMessage" {
            if !text(&item, "text").is_empty() {
                self.last_message = text(&item, "text").to_owned();
            }
            if kind == "item.completed"
                && let Some(questions) = item["questions"].as_array()
            {
                let fields = questions
                    .iter()
                    .enumerate()
                    .map(|(i, q)| {
                        json!({
                        "id":i.to_string(),"title":q["title"],"options":q["options"].as_array().into_iter().flatten().map(|label|json!({
                        "label":label}
                        )).collect::<Vec<_>>()}
                        )
                    })
                    .collect::<Vec<_>>();
                if let Ok(fields) = parse("questions", fields.into()) {
                    self.emit(json!({
                    "type":"chat.question","question":{
                    "id":hex_digest(&format!("{}:{}",self.thread,text(&item,"id"))),"blocking":false,"fields":fields}
                    }
                    ))
                    .await?;
                }
            }
        }
        self.emit(json!({
        "type":kind,"item":chat_item(item)}
        ))
        .await
    }
    async fn incoming(&mut self, incoming: Incoming) -> Result<()> {
        let params = incoming.params;
        if let Some(request_id) = incoming.id {
            if incoming.method == "item/tool/requestUserInput"
                && !text(&params, "itemId").is_empty()
                && (self.thread.is_empty() || params["threadId"] == self.thread)
            {
                let fields = params["questions"]
                    .as_array()
                    .into_iter()
                    .flatten()
                    .map(|q| {
                        json!({
                        "id":q["id"],"title":q["question"],"secret":q["isSecret"].as_bool().unwrap_or(false),"options":q["options"].as_array().cloned().unwrap_or_default()}
                        )
                    })
                    .collect::<Vec<_>>();
                if let Ok(fields) = parse("questions", fields.into()) {
                    let thread = if self.thread.is_empty() {
                        text(&params, "threadId")
                    } else {
                        &self.thread
                    };
                    let id = hex_digest(&format!("{thread}:{}", text(&params, "itemId")));
                    self.questions.insert(
                        id.clone(),
                        Question {
                            request_id,
                            message_id: None,
                        },
                    );
                    return self
                        .emit(json!({
                        "type":"chat.question","question":{
                        "id":id,"blocking":params["isBlocking"]!=false,"fields":fields}
                        }
                        ))
                        .await;
                }
            }
            return self.session.rpc.reject(request_id).await;
        }
        if params["threadId"].is_string()
            && !self.thread.is_empty()
            && params["threadId"] != self.thread
        {
            return Ok(());
        }
        match incoming.method.as_str() {
            "serverRequest/resolved" => {
                let ids = self
                    .questions
                    .iter()
                    .filter(|(_, q)| q.request_id == params["requestId"])
                    .map(|(id, _)| id.clone())
                    .collect::<Vec<_>>();
                for id in ids {
                    let question = self.questions.remove(&id).unwrap();
                    if let Some(message) = question.message_id {
                        self.acknowledge(&message).await?;
                    }
                    self.emit(json!({
                    "type":"chat.question.closed","questionId":id}
                    ))
                    .await?;
                }
            }
            "turn/started" => {
                self.turn = text(&params["turn"], "id").to_owned();
                self.emit(json!({
                "type":"turn.started"}
                ))
                .await?;
            }
            "item/started" | "item/completed" => {
                self.item(params["item"].clone(), &incoming.method.replace('/', "."))
                    .await?
            }
            "item/agentMessage/delta" => {
                let value = self
                    .texts
                    .entry(text(&params, "itemId").to_owned())
                    .or_default();
                if value.len() + text(&params, "delta").len() > 5_000_000 {
                    return Err(Error::new(
                        502,
                        "Conversation output exceeded the supported limit.",
                    ));
                }
                value.push_str(text(&params, "delta"));
                let value = value.clone();
                self.emit(json!({
                "type":"item.updated","item":{
                "id":params["itemId"],"type":"agent_message","text":value}
                }
                ))
                .await?;
            }
            "turn/completed" => self.completed = Some(params["turn"].clone()),
            _ => {}
        }
        Ok(())
    }
    async fn request(&mut self, method: &str, params: Value) -> Result<Value> {
        let rpc = self.session.rpc.clone();
        let request = rpc.request(method, params);
        tokio::pin!(request);
        loop {
            tokio::select! {
                            _=self.cancel.cancelled()=>return Err(Error::new(409,"Conversation stopped.")),
                            result=&mut request=>return result,
                            incoming=self.session.incoming.recv()=>{
            let incoming=incoming.ok_or_else(||Error::new(503,"Codex disconnected before finishing the response. Resume the conversation to continue."))?;
            self.incoming(incoming).await?;
            }
                        }
        }
    }
    async fn execute(&mut self, plan: &Value) -> Result<()> {
        let mut settings = json!({
        "cwd":plan["cwd"],"approvalPolicy":"never","sandbox":if plan["sandbox"]=="yolo"{
        "danger-full-access"}
        else{
        text(plan,"sandbox")}
        ,"developerInstructions":plan["instructions"],"config":{
        "features.default_mode_request_user_input":true,"model_reasoning_effort":plan["reasoning"],"sandbox_workspace_write":{
        "network_access":true,"writable_roots":plan["writableRoots"]}
        }
        }
        );
        if !text(plan, "model").is_empty() {
            settings["model"] = plan["model"].clone();
        }
        let resume = plan["sessionId"].is_string();
        if resume {
            settings["threadId"] = plan["sessionId"].clone();
            settings["excludeTurns"] = false.into();
        }
        let result = self
            .request(
                if resume {
                    "thread/resume"
                } else {
                    "thread/start"
                },
                settings,
            )
            .await?;
        self.thread = text(&result["thread"], "id").to_owned();
        if self.thread.is_empty() {
            return Err(Error::new(502, "Codex returned an invalid conversation."));
        }
        self.emit(json!({
        "type":"chat.question.closed"}
        ))
        .await?;
        self.emit(json!({
        "type":"thread.started","thread_id":self.thread}
        ))
        .await?;
        let mut turns = result["thread"]["turns"]
            .as_array()
            .cloned()
            .unwrap_or_default();
        let paginated = result["thread"]["historyMode"] == "paginated";
        if resume && paginated {
            turns.clear();
            let mut cursor = Value::Null;
            let mut cursors = HashSet::new();
            loop {
                let mut params = json!({
                "threadId":self.thread,"limit":100,"itemsView":"full","sortDirection":"desc"}
                );
                if !cursor.is_null() {
                    params["cursor"] = cursor;
                }
                let page = self.request("thread/turns/list", params).await?;
                turns.extend(page["data"].as_array().cloned().ok_or_else(|| {
                    Error::new(502, "Codex returned invalid conversation history.")
                })?);
                cursor = page["nextCursor"].clone();
                if cursor.is_null() {
                    break;
                }
                if !cursors.insert(cursor.to_string()) || cursors.len() > 1000 {
                    return Err(Error::new(
                        502,
                        "Codex returned invalid conversation pagination.",
                    ));
                }
            }
        }
        for turn in &turns {
            for item in turn["items"].as_array().into_iter().flatten() {
                if item["type"] == "userMessage"
                    && let Some(id) = item["clientId"].as_str()
                {
                    self.acknowledge(id).await?;
                }
            }
        }
        let accepted = turns.iter().rev().find(|t| {
            t["items"].as_array().is_some_and(|items| {
                items
                    .iter()
                    .any(|i| i["clientId"] == plan["execution"]["messageId"])
            })
        });
        let previous = if accepted.is_some() && plan["execution"]["recovery"] == true {
            if paginated {
                turns.first()
            } else {
                turns.last()
            }
        } else {
            accepted
        };
        if let Some(previous) = previous {
            self.acknowledge(text(&plan["execution"], "messageId"))
                .await?;
            if previous["status"] == "completed" {
                for item in previous["items"].as_array().into_iter().flatten() {
                    self.item(item.clone(), "item.completed").await?;
                }
                return self.finish(plan).await;
            }
        }
        let message = if previous.is_some() {
            format!(
                "Continue the interrupted conversation from its last completed step. Preserve completed work and verify external effects before repeating any action. The pending user request is:\n{}",
                text(&plan["execution"], "text")
            )
        } else {
            text(&plan["execution"], "text").to_owned()
        };
        let mut params = json!({
        "threadId":self.thread,"input":crate::attachments::input(&message, &plan["execution"]["attachments"], Path::new(text(plan,"inputDirectory"))),"effort":plan["reasoning"]}
        );
        if previous.is_none() {
            params["clientUserMessageId"] = plan["execution"]["messageId"].clone();
        }
        if !text(plan, "model").is_empty() {
            params["model"] = plan["model"].clone();
        }
        let started = self.request("turn/start", params).await?;
        self.turn = text(&started["turn"], "id").to_owned();
        self.acknowledge(text(&plan["execution"], "messageId"))
            .await?;
        let mut interval = tokio::time::interval(Duration::from_millis(250));
        interval.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Skip);
        while self.completed.is_none() {
            tokio::select! {
                            _=self.cancel.cancelled()=>return Err(Error::new(409,"Conversation stopped.")),
                            incoming=self.session.incoming.recv()=>{
            let incoming=incoming.ok_or_else(||Error::new(503,"Codex disconnected before finishing the response. Resume the conversation to continue."))?;
            self.incoming(incoming).await?;
            }
            ,
                            _=interval.tick()=>self.steer(plan).await?,
                        }
        }
        let turn = self.completed.take().unwrap();
        if turn["status"] != "completed" {
            self.emit(json!({
            "type":"turn.failed","error":turn.get("error").cloned().unwrap_or_else(||json!({
            "message":"Conversation interrupted."}
            ))}
            ))
            .await?;
            return Err(Error::new(409, "Conversation interrupted."));
        }
        self.finish(plan).await
    }
    async fn steer(&mut self, plan: &Value) -> Result<()> {
        let Ok(bytes) =
            tokio::fs::read(Path::new(text(plan, "inputDirectory")).join("messages.json")).await
        else {
            return Ok(());
        };
        if bytes.len() > 2_000_000 {
            return Err(Error::bad("Chat inbox is too large."));
        }
        let Ok(messages) = serde_json::from_slice::<Vec<Value>>(&bytes) else {
            return Ok(());
        };
        for message in messages {
            let id = text(&message, "id");
            if self.completed.is_some()
                || self.seen.contains(id)
                || !self.attempted.insert(id.to_owned())
            {
                continue;
            }
            if let Some(question) = self.questions.get_mut(text(&message, "questionId"))
                && let Some(answers) = message["answers"].as_object()
            {
                question.message_id = Some(id.to_owned());
                let result = json!({
                "answers":answers.iter().map(|(id,answers)|(id.clone(),json!({
                "answers":answers}
                ))).collect::<serde_json::Map<_,_>>()}
                );
                self.session
                    .rpc
                    .reply(question.request_id.clone(), result)
                    .await?;
                continue;
            }
            let params = json!({
            "threadId":self.thread,"expectedTurnId":self.turn,"clientUserMessageId":id,"input":crate::attachments::input(text(&message,"text"), &message["attachments"], Path::new(text(plan,"inputDirectory")))}
            );
            if self.request("turn/steer", params).await.is_err() {
                break;
            }
            self.acknowledge(id).await?;
        }
        Ok(())
    }
    async fn finish(&self, plan: &Value) -> Result<()> {
        atomic_write(
            Path::new(text(plan, "output")),
            self.last_message.as_bytes(),
        )
        .await?;
        self.emit(json!({
        "type":"turn.completed"}
        ))
        .await
    }
}
pub async fn run(
    config: &Config,
    home: &Path,
    plan: Value,
    events: mpsc::Sender<Value>,
    cancel: CancellationToken,
) -> Result<()> {
    let args = plan["args"]
        .as_array()
        .into_iter()
        .flatten()
        .filter_map(Value::as_str)
        .map(str::to_owned)
        .collect::<Vec<_>>();
    let session = Session::codex(config, home, &args, Some(Path::new(text(&plan, "cwd")))).await?;
    let mut chat = Chat {
        session,
        events,
        cancel,
        thread: String::new(),
        turn: String::new(),
        completed: None,
        last_message: String::new(),
        seen: HashSet::new(),
        attempted: HashSet::new(),
        texts: HashMap::new(),
        questions: HashMap::new(),
    };
    let result = chat.execute(&plan).await;
    chat.session.close().await;
    result
}
